src/Logary/Metric/Gauge.fs

```diff
@@ -0,0 +1,79 @@
+namespace Logary.Metric
+
+
+type GaugeConf =
+  {
+   basicInfo: BasicConf
+   histogramConf: HistogramConf option
+  }
+
+  interface MetricBuilder<IGauge> with
+    member x.build labels registry =
+      let histrgomMetric =
+        x.histogramConf
+        |> Option.map (fun conf ->
+          let labelValues = conf.basicInfo.labelNames |> Array.map (fun name -> labels.[name])
+          conf |> registry.registerMetric |> Metric.labels labelValues)
+      new Gauge(x, labels, histrgomMetric) :> IGauge
+
+    member x.basicConf = x.basicInfo
+
+  static member create (name, description) =
+    let basic = { name =  name; description = description; labelNames = [||] }
+    {basicInfo = basic; histogramConf = None}
+
+  static member create (basic) =
+    {basicInfo = basic; histogramConf = None}
+
+and Gauge(conf, labels, histogram) =
+
+  let mutable gaugeValue = new DoubleAdder()
+  let changeGaugeValueLock = obj ()
+
+  let tryDoHistogramObserve () =
+    histogram |> Option.iter(fun histogram ->
+      let sum = gaugeValue.Sum ()
+      histogram.observe sum)
+
+  interface IGauge with
+    member x.inc value =
+      match histogram with
+      | Some histogram ->
+        lock changeGaugeValueLock <| fun () ->
+          gaugeValue.Add value
+          gaugeValue.Sum () |> histogram.observe
+      | None -> gaugeValue.Add value
+
+    member x.dec value =
+      (x :> IGauge).inc -value
+
+    member x.set value =
```

Comment:
Should be decouple gauge from counter, perhaps?

Reply:
this is what i was wondering,  according to this:

- https://prometheus.io/docs/instrumenting/writing_clientlibs/#counter
- https://godoc.org/github.com/prometheus/client_golang/prometheus#Counter

> Counter is a Metric that represents a single numerical value that only ever goes up. That implies that it cannot be used to count items whose number can also go down, e.g. the number of currently running goroutines. Those "counters" are represented by Gauges.

>A Counter is typically used to count requests served, tasks completed, errors occurred, etc. 

i don't really know if a `counter` is different with `gauge` at prometheus's under implementation, or prometheus treat them the same thing.

if they are only different at the semantic level, i think we should not expose a counter, if user only use gauge's `inc` method, it act like a counter , if user change their mind to `dec` something, then it act like a gauge.

Reply:
but if prometheus treat them as different, then a `counter` can be simply wrapped by a interface only expose `inc` method, and check the value to be positive.